panic in `(auth *Authenticator) handshakeFirstTime`

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

It's an alert from internal platform

### 2. What did you expect to see? (Required)

No error.

### 3. What did you see instead (Required)

```
2025-10-22T10:34:04.181594410Z;ip-10-0-11-66.us-west-2.compute.internal;tidb1625120103073681;db-10691416346745616064-tiproxy-qrowrs;tiproxy [2025/10/22 10:34:04.181 +00:00] [ERROR] [main.proxy] [waitgroup/waitgroup.go:64] [panic in a recoverable goroutine]  [err="runtime error: slice bounds out of range [:4] with capacity 1"] ["stack trace"="github.com/pingcap/tiproxy/lib/util/waitgroup.recoverFromErr\n\t/workspace/source/tiproxy/lib/util/waitgroup/waitgroup.go:66\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:791\nruntime.goPanicSliceAcap\n\t/usr/local/go/src/runtime/panic.go:141\ngithub.com/pingcap/tiproxy/pkg/proxy/net.ParseSSLRequestOrHandshakeResp\n\t/workspace/source/tiproxy/pkg/proxy/net/mysql.go:50\ngithub.com/pingcap/tiproxy/pkg/proxy/backend.(*Authenticator).handshakeFirstTime\n\t/workspace/source/tiproxy/pkg/proxy/backend/authenticator.go:111\ngithub.com/pingcap/tiproxy/pkg/proxy/backend.(*BackendConnManager).Connect\n\t/workspace/source/tiproxy/pkg/proxy/backend/backend_conn_mgr.go:213\ngithub.com/pingcap/tiproxy/pkg/proxy/client.(*ClientConnection).Run\n\t/workspace/source/tiproxy/pkg/proxy/client/client_conn.go:50\ngithub.com/pingcap/tiproxy/pkg/proxy.(*SQLServer).onConn\n\t/workspace/source/tiproxy/pkg/proxy/proxy.go:197\ngithub.com/pingcap/tiproxy/pkg/proxy.(*SQLServer).Run.func2.1\n\t/workspace/source/tiproxy/pkg/proxy/proxy.go:137\ngithub.com/pingcap/tiproxy/lib/util/waitgroup.(*WaitGroup).RunWithRecover.func1\n\t/workspace/source/tiproxy/lib/util/waitgroup/waitgroup.go:53"]
2025-10-22T10:34:00.700417011Z;ip-10-0-72-140.us-west-2.compute.internal;tidb1625120103073681;db-10691416346745616064-tiproxy-l9xlpl;tiproxy [2025/10/22 10:34:00.700 +00:00] [ERROR] [main.proxy] [waitgroup/waitgroup.go:64] [panic in a recoverable goroutine]  [err="runtime error: slice bounds out of range [:4] with capacity 3"] ["stack trace"="github.com/pingcap/tiproxy/lib/util/waitgroup.recoverFromErr\n\t/workspace/source/tiproxy/lib/util/waitgroup/waitgroup.go:66\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:791\nruntime.goPanicSliceAcap\n\t/usr/local/go/src/runtime/panic.go:141\ngithub.com/pingcap/tiproxy/pkg/proxy/net.ParseSSLRequestOrHandshakeResp\n\t/workspace/source/tiproxy/pkg/proxy/net/mysql.go:50\ngithub.com/pingcap/tiproxy/pkg/proxy/backend.(*Authenticator).handshakeFirstTime\n\t/workspace/source/tiproxy/pkg/proxy/backend/authenticator.go:111\ngithub.com/pingcap/tiproxy/pkg/proxy/backend.(*BackendConnManager).Connect\n\t/workspace/source/tiproxy/pkg/proxy/backend/backend_conn_mgr.go:213\ngithub.com/pingcap/tiproxy/pkg/proxy/client.(*ClientConnection).Run\n\t/workspace/source/tiproxy/pkg/proxy/client/client_conn.go:50\ngithub.com/pingcap/tiproxy/pkg/proxy.(*SQLServer).onConn\n\t/workspace/source/tiproxy/pkg/proxy/proxy.go:197\ngithub.com/pingcap/tiproxy/pkg/proxy.(*SQLServer).Run.func2.1\n\t/workspace/source/tiproxy/pkg/proxy/proxy.go:137\ngithub.com/pingcap/tiproxy/lib/util/waitgroup.(*WaitGroup).RunWithRecover.func1\n\t/workspace/source/tiproxy/lib/util/waitgroup/waitgroup.go:53"]
```

### 4. What is your version? (Required)

<!-- Paste the output of tiproxy version -->
<!-- Paste the output of tiproxyctl version if related -->
<!-- Paste the output of SELECT tidb_version()  if related -->

--- a/pkg/proxy/backend/authenticator.go
+++ b/pkg/proxy/backend/authenticator.go
@@ -108,6 +108,13 @@ func (auth *Authenticator) handshakeFirstTime(ctx context.Context, logger *zap.L
 	if err != nil {
 		return err
 	}
+	// The first packet is either `HandshakeResponse` or `SSLRequest`. They are both at least 32 bytes.
+	// Ref https://dev.mysql.com/doc/dev/mysql-server/8.0.44/page_protocol_connection_phase_packets_protocol_handshake_response.html
+	// Ref https://dev.mysql.com/doc/dev/mysql-server/8.0.44/page_protocol_connection_phase_packets_protocol_ssl_request.html
+	// The following logic will panic if the length is less than 32, so we check it here.
+	if len(pkt) < 32 {
+		return errors.Wrap(ErrClientHandshake, mysql.ErrMalformPacket)
+	}
 	isSSL := pnet.ParseSSLRequestOrHandshakeResp(pkt)
 	frontendCapability := pnet.Capability(binary.LittleEndian.Uint32(pkt))
 	if isSSL {

--- a/pkg/proxy/backend/authenticator_test.go
+++ b/pkg/proxy/backend/authenticator_test.go
@@ -4,6 +4,7 @@
 package backend
 
 import (
+	"encoding/binary"
 	"strings"
 	"testing"
 
@@ -595,4 +596,29 @@ func TestUpgradeBackendCap(t *testing.T) {
 		})
 		clean()
 	}
+}
+
+func TestMalformedHandshakePacket(t *testing.T) {
+	tc := newTCPConnSuite(t)
+	ts, clean := newTestSuite(t, tc)
+
+	customClientRunner := func(packetIO pnet.PacketIO) error {
+		_, err := packetIO.ReadPacket()
+		if err != nil {
+			return err
+		}
+
+		malformedPacket := make([]byte, 31)
+		binary.LittleEndian.PutUint32(malformedPacket[0:4], ts.mc.capability.Uint32())
+
+		return packetIO.WritePacket(malformedPacket, true)
+	}
+
+	ts.runAndCheck(t, func(t *testing.T, ts *testSuite) {
+		require.Error(t, ts.mp.err)
+		require.Contains(t, ts.mp.err.Error(), "Malform packet error: Fails to handshake with the client")
+		require.Equal(t, SrcProxyMalformed, Error2Source(ts.mp.err))
+	}, customClientRunner, ts.mb.authenticate, ts.mp.authenticateFirstTime)
+
+	clean()
 }

